src/cluster/agglomerative.rs

```diff
@@ -0,0 +1,219 @@
+  pub fn fit<M: Matrix<T>>(
+    data: &M,
+    k: usize,
+  ) -> Result<ClusterResult<T>, Failed>;
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct PrimitiveClustering {}
+
+impl PrimitiveClustering {
+  fn compute(&self) ->  {
+    // 1: procedure Primitive_clustering(S, d) . S: node labels, d: pairwise dissimilarities
+    // 2: N ← |S| . Number of input nodes
+    // 3: L ← [ ] . Output list
+    // 4: size[x] ← 1 for all x ∈ S
+    // 5: for i ← 0, . . . , N − 2 do
+    // 6: (a, b) ← argmin(S×S)\∆ d
+    // 7: Append (a, b, d[a, b]) to L.
+    // 8: S ← S \ {a, b}
+    // 9: Create a new node label n /∈ S.
+    // 10: Update d with the information
+    // d[n, x] = d[x, n] = Formula(d[a, x], d[b, x], d[a, b], size[a], size[b], size[x])
+    // for all x ∈ S.
+    // 11: size[n] ← size[a] + size[b]
+    // 12: S ← S ∪ {n}
+    // 13: end for
+    // 14: return L . the stepwise dendrogram, an ((N − 1) × 3)-matrix
+    // 15: end procedure
+    // (As usual, ∆ denotes the diagonal in the Cartesian product S × S.)
+  }
+}
+
+///
+/// An implementation of `PRIMITIVE_CLUSTERING`
+///
+impl SAHNClustering for PrimitiveClustering {
+
+}
+
+
+///
+/// The triple defined as (observation_1, observation_2, distance).
+/// This defines the type for the elements of the list that 
+///   represents the Stepwise Dendrogram (Müllner, 2011)
+/// 
+#[derive(Debug)]
+struct PairwiseDissimilarity<'a> {
+    node1: Vec<RealNumber>;
+    node2: Vec<RealNumber>;
+    distance: f32;
+    position: usize;
+};
+
+///
+/// Compare distance
+/// 
+impl Ord for PairwiseDissimilarity {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.distance.cmp(&other.distance)
+    }
+}
+
+impl PartialOrd for PairwiseDissimilarity {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl PartialEq for PairwiseDissimilarity {
+    fn eq(&self, other: &Self) -> bool {
+        self.distance == other.distance
+    }
+}
+
+///
+/// Distance update formulas
+/// Formula for d(I ∪ J, K)  as in Fig.2 in Müllner, 2011
+/// 
+impl PairwiseDissimilarity<'a> {
+  fn single_euclidian(&self, I: &Vec<f32>, J: &Vec<f32>, K: &Vec<f32>) {
+    // min(d(I, K), d(J, K))
+    cmp::min(Euclidian {}.distance(I, K), Euclidian {}.distance(J, K))
+  }
+
+  fn complete_euclidian(&self, I: &Vec<f32>, J: &Vec<f32>, K: &Vec<f32>) {
+    // max(d(I, K), d(J, K))
+    cmp::max(Euclidian {}.distance(I, K), Euclidian {}.distance(J, K))
+  }
+
+  // ...
+}
+
+///
+/// Struct to hold result of linkage and clustering
+/// 
+/// > The term 'dendrogram' has been used with three different
+/// > meanings: a mathematical object, a data structure and
+/// > a graphical representation of the former two. In the course of this section, we
+/// > define a data structure and call it 'stepwise dendrogram'.
+/// 
+/// Rust implementation: see options at <https://stackoverflow.com/a/40897053/2536357>
+pub struct ClusterResult<'a> {
```

Comment:
I thought the result will be cluster labels. Btw, we do not need method `predict` because we cannot retrieve the cluster membership of new data. 
Linkage matrix of the size (N-1)x3 (or (N-1)x2 if we use row number as an index) is also an important result but it is not as important as cluster labels. I think we do need something similar to `children_` attribute of the [AgglomerativeClustering](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.AgglomerativeClustering.html) because one of the cool features of the algorithm is that you can use dendrogram to decide how many clusters you need.